src/trait_/common_ord.rs

```diff
@@ -214,79 +231,131 @@ pub fn build_ord_signature(
-fn build_recursive_order(
-	trait_: &DeriveTrait,
+fn build_discriminant_order(
+	repr: Option<Representation>,
+	item: &Item,
+	generics: &SplitGenerics<'_>,
 	variants: &[Data<'_>],
-	incomparable: &TokenStream,
+	path: &Path,
+	method: &TokenStream,
 ) -> TokenStream {
-	let mut less = quote! { ::core::cmp::Ordering::Less };
-	let mut greater = quote! { ::core::cmp::Ordering::Greater };
+	use std::{borrow::Cow, ops::Deref};
+
+	use proc_macro2::Span;
+	use syn::{parse_quote, Expr, ExprLit, Lit, LitInt};
+
+	let mut discriminants = Vec::<Cow<Expr>>::with_capacity(variants.len());
+	let mut has_non_isize = false;
+
+	for variant in variants {
+		let discriminant = if let Some(discriminant) = variant.discriminant {
+			if !has_non_isize
+				&& !matches!(
+					discriminant,
+					Expr::Lit(ExprLit {
+						lit: Lit::Int(_),
+						..
+					})
+				) {
+				has_non_isize = true;
+			}
 
-	// Add `Option` to `Ordering` if we are implementing `PartialOrd`.
-	if let DeriveTrait::PartialOrd = trait_ {
-		less = quote! { ::core::option::Option::Some(#less) };
-		greater = quote! { ::core::option::Option::Some(#greater) };
-	}
+			Cow::Borrowed(discriminant)
+		} else if let Some(discriminant) = discriminants.last().map(Deref::deref) {
+			let discriminant = if let Expr::Lit(ExprLit {
+				lit: Lit::Int(int), ..
+			}) = discriminant
+			{
+				let int = if let Ok(int) = int.base10_parse::<i128>() {
+					let int = int + 1;
 
-	let mut different = Vec::with_capacity(variants.len());
-	let variants: Vec<_> = variants.iter().filter(|v| !v.is_incomparable()).collect();
+					if !has_non_isize {
+						if let Ok(max) = i128::try_from(isize::MAX) {
+							if int > max {
+								has_non_isize = true;
+							}
+						}
+					}
 
-	// Build separate `match` arms to compare different variants to each
-	// other. The index for these variants is used to determine which
-	// `Ordering` to return.
-	for (index, variant) in variants.iter().enumerate() {
-		let pattern = &variant.self_pattern();
+					int.to_string()
+				} else if let Ok(int) = int.base10_parse::<u128>() {
+					let int = int + 1;
 
-		// The first variant is always `Less` then everything.
-		if index == 0 {
-			different.push(quote! {
-				#pattern => #less,
-			})
-		}
-		// The last variant is always `Greater` then everything.
-		else if index == variants.len() - 1 {
-			different.push(quote! {
-				#pattern => #greater,
-			})
-		}
-		// Any variant between the first and last.
-		else {
-			// Collect all variants that are `Less`.
-			let cases = variants
-				.iter()
-				.enumerate()
-				.filter(|(index_other, _)| *index_other < index)
-				.map(|(_, variant_other)| variant_other.other_pattern_skip().clone())
-				.collect();
-
-			// Build one match arm pattern with all variants that are `Greater`.
-			let pattern_less = PatOr {
-				attrs: Vec::new(),
-				leading_vert: None,
-				cases,
+					if !has_non_isize {
+						if let Ok(max) = u128::try_from(isize::MAX) {
```

Comment:
errr... what would happen if an isize doesn't fit a u128, I feel like this would be a place for an expect instead of silently ignoring

Reply:
I'm not exactly ignoring it, but I assume that it's not an `isize`, which is fine.

But actually I just realized that this whole `isize` check can't work, because I'm not using the target architecture `isize` but the build architecture one.

I will remove it completely ...

Reply:
Doesn't it say somewhere in the specification the minimum target architecture for Rust has to be 16-bit? I could check for that.

Reply:
Took target architecture pointer width into account.